django/db/models/query.py

```diff
@@ -1283,7 +1286,11 @@ def exists(self):
         Return True if the QuerySet would have any results, False otherwise.
         """
         if self._result_cache is None:
+            if getattr(self, "_many_related_manager", None):
+                return self._many_related_manager.exists()
+
             return self.query.has_results(using=self.db)
+
```

Comment:
This bi-directionnaly couples the SQL resolving layer with fields managers which I don't think we want to do here.

As mentioned on the ticket I don't think we should venture down the path of chaining optimization as it's quite hard to get right.

For example the code you have added here doesn't take into account filtering, annotation, or any other things that a query might perform from the point where it originates from the related manager to when `count` is called.

The check for `self._result_cache is None` is naive and can be demonstrated to break by simply doing something like `self.article.authors.exclude(username="Jean").count()`.

Reply:
Should I reverse the changes back to commit [417c12e](https://github.com/django/django/pull/17811/commits/417c12e1af87f4bac98eb4c4c69735e53c84707d)?

Is there a recommended way to do this with git? I was looking into the two commands:
- `git reset --hard 417c12e; git push origin HEAD --force`
- `git revert 417c12e; git push origin HEAD`